### ISSUE TITLE
Add Dropout layer

### ISSUE THREAD

**Issue by leonardoaraujosantos** [leonardoaraujosantos/DLMatFramework#2]
# Introduction
Add dropout layer, to improve generalization

### Task
1. Create a new class from BaseLayer
2. Add a new test to check if forward/backward prop are working

#### Training
While training, dropout is implemented by only keeping a neuron active with some probability p (a hyperparameter), or setting it to zero otherwise.
### Python code for inverted dropout (Forward propagation)
``` python
""" 
Inverted Dropout: Recommended implementation example.
We drop and scale at train time and don't do anything at test time.
"""

p = 0.5 # probability of keeping a unit active. higher = less dropout

def train_step(X):
  # forward pass for example 3-layer neural network
  H1 = np.maximum(0, np.dot(W1, X) + b1)
  U1 = (np.random.rand(*H1.shape) < p) / p # first dropout mask. Notice /p!
  H1 *= U1 # drop!
  H2 = np.maximum(0, np.dot(W2, H1) + b2)
  U2 = (np.random.rand(*H2.shape) < p) / p # second dropout mask. Notice /p!
  H2 *= U2 # drop!
  out = np.dot(W3, H2) + b3
  
  # backward pass: compute gradients... (not shown)
  # perform parameter update... (not shown)
  
def predict(X):
  # ensembled forward pass
  H1 = np.maximum(0, np.dot(W1, X) + b1) # no scaling necessary
  H2 = np.maximum(0, np.dot(W2, H1) + b2)
  out = np.dot(W3, H2) + b3
```

## Reference
* [Dropout forward](https://github.com/leonardoaraujosantos/DLMatFramework/blob/master/learn/cs231n/assignment2/cs231n/layers.py#L365)
* [Dropout backward](https://github.com/leonardoaraujosantos/DLMatFramework/blob/master/learn/cs231n/assignment2/cs231n/layers.py#L426)
* [Cs231n Notes](http://cs231n.github.io/neural-networks-2/#reg)
* [AI Book](https://leonardoaraujosantos.gitbooks.io/artificial-inteligence/content/dropout_layer.html)

**Comment by leonardoaraujosantos** [comment]
## Current Results

### Without Dropout
![nodropout](https://cloud.githubusercontent.com/assets/898383/22335074/1cae6880-e3d5-11e6-9ba5-b12de53d5148.png)

### With dropout (0.5)
![with_dropout](https://cloud.githubusercontent.com/assets/898383/22335083/26dca04c-e3d5-11e6-8095-3ca6424c9839.png)

### With dropout (0.2)
![dropout_02](https://cloud.githubusercontent.com/assets/898383/22335294/0c658dae-e3d6-11e6-9179-7b4eea74d76b.png)
